test/parallel/test-http-upgrade-server-with-body.js

```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+const common = require('../common');
+const assert = require('assert');
+const http = require('http');
+const net = require('net');
+
+const server = http.createServer();
+
+server.on('request', common.mustNotCall());
+
+server.on('upgrade', function(req, socket, upgradeHead) {
+  let reqBody = '';
+  req.on('data', common.mustCall((str) => { reqBody += str; }));
```

Comment:
What happens if all the incoming data cannot be read before the socket is freed?

Reply:
Hmm, this is a good question. I'm not actually sure how to test this though! Any pointers? I'll do some digging and see if I can work this out but if you have a suggestion for how to check that would be useful.

Reply:
Is the `'upgrade'` event after receiving the headers? If so, it would be sufficient to send the request body or part of it after the server emits `'upgrade'`, no?

Reply:
Ah, gotcha. I had misunderstood I think. I've just tested and it does look like this is a problem. I'll look into it and update here once I have a fix, and convert to draft in the meantime. Good catch!